src/ceph-volume/ceph_volume/util/encryption.py

```diff
@@ -23,11 +24,56 @@ def get_key_size_from_conf():
 
     if key_size not in ['256', '512']:
         logger.warning(("Invalid value set for osd_dmcrypt_key_size ({}). "
-                        "Falling back to {}bits".format(key_size, default_key_size)))
+                        "Falling back to {} bits".format(key_size, default_key_size)))
         return default_key_size
 
     return key_size
 
+
+def execCmd(cmdline):
+    cmd = subprocess.Popen(cmdline, shell=True, stdout=subprocess.PIPE)
+    stdout, _ = cmd.communicate()
+    return stdout.decode("utf-8")
+
+
+def bypass_workqueues(device):
+    """
+    1) Check if cryptsetup has support for no_read_workqueue.
+    If so, it also supports no_write_workqueue.
+    cryptsetup has support for this since 2.3.4
+
+    2) Also check if we are using a flash (non-rotational) device
+    or not. I.e. Spinning media might still benefit from queuing.
+
+    Only if both 1) and 2) are true we do _not_ want to use work queues
+    Note:
+    crypt version 1.22 and higher have support for this. Example:
+    modprobe dm-crypt
+    dmsetup targets
+
+    integrity        v1.10.0
+    crypt            v1.23.0
+    striped          v1.6.0
+    linear           v1.4.0
+    error            v1.5.0
+
+    If dm-crypt does not have support for no_read_workqueue /
+    no_write_workqueue the options will be discarded and
+    logged by the linux kernel:
+
+    device-mapper: table: major:minor: crypt: Invalid feature arguments
+    device-mapper: ioctl: error adding target to table
+
+    The encrypted device will come online without the options active.
+    """
+    cryptsetup_help = execCmd("cryptsetup --help 2>/dev/null")
+
+    if '--perf-no_read_workqueue' not in cryptsetup_help or is_rotational(device):
```

Comment:
```suggestion
    if '--perf-no_read_workqueue' not in cryptsetup_help or lsblk(device)['ROTA'] == '1':
```

As far as I know, they key 'ROTA' is always present but we might need to double-check that. If it can be missing, then maybe we would need a try/except KeyError sequence.

Reply:
@guits Thanks for reviewing this. I'll make time to address the issues where possible and see if we can make it so that it is fit for merging.